instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/LambdaImpl.java

```diff
@@ -0,0 +1,141 @@
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.services.lambda.LambdaClient;
+import software.amazon.awssdk.services.lambda.model.InvokeRequest;
+
+// this class is only used from LambdaAccess from method with @NoMuzzle annotation
+final class LambdaImpl {
+
+  private static final Logger logger = LoggerFactory.getLogger(LambdaImpl.class);
+
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  static {
+    // Force loading of LambdaClient; this ensures that an exception is thrown at this point when the
+    // Lambda library is not present, which will cause LambdaAccess to have enabled=false in library mode.
+    @SuppressWarnings("unused")
+    String ensureLoadedDummy = LambdaClient.class.getName();
+  }
+
+  private LambdaImpl() {}
+
+  @Nullable
+  static SdkRequest modifyRequest(
+      SdkRequest request, Context otelContext, TextMapPropagator messagingPropagator) {
+    System.out.println(
+        "LambdaImpl.modifyRequest " + request.getClass().getCanonicalName() + " from "
+            + request.getClass().getClassLoader() + "messagingPropagator" + messagingPropagator);
+
+    if (messagingPropagator == null) {
+      return null;
+    }
+
+    if (request instanceof InvokeRequest) {
+      System.out.println("injectIntoInvokeRequest");
+      return injectIntoInvokeRequest((InvokeRequest) request, otelContext, messagingPropagator);
+    } else {
+      return null;
+    }
+  }
+
+  private static SdkRequest injectIntoInvokeRequest(
+      InvokeRequest request, Context otelContext, TextMapPropagator messagingPropagator) {
+    String modifiedClientContext = injectIntoClientContext(request.clientContext(), otelContext,
+        messagingPropagator);
+    return request.toBuilder().clientContext(modifiedClientContext).build();
+  }
+
+  private static String injectIntoClientContext(
+      String originalClientContext,
+      Context otelContext,
+      TextMapPropagator messagingPropagator) {
+
+    try {
+      Map<String, Object> clientContext = deserialiseClientContext(originalClientContext);
+      if (clientContext == null) {
+        clientContext = new HashMap<>();
+      }
+
+      Object customObject = clientContext.computeIfAbsent("custom",
+          k -> new HashMap<String, String>());
+      if (!(customObject instanceof Map)) {
+        return originalClientContext;
+      }
+
+      @SuppressWarnings("unchecked")
+      Map<String, String> custom = (Map<String, String>) customObject;
+      messagingPropagator.inject(
+          otelContext,
+          custom,
+          MapSetter.INSTANCE
+      );
+      String modifiedClientContext = serialiseClientContext(clientContext);
+      // Make sure we don't exceed the size limit imposed by AWS https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html
+      if (modifiedClientContext.length() <= 3583) {
```

Comment:
Do you know if 3583 is the limit before of after base64 encoding? (I would read "Up to 3,583 bytes of base64-encoded data" as 3583 bytes before encoding, but it can be either way).

Reply:
I tried to learn that but couldn't find any reliable source. Applying the limit after base64 encoding is more conservative (increases the risk that we will unnecessarily give up on passing the context but reduces the risk that we will crash something with our instrumentation)